src/Core/src/Platform/iOS/WindowViewController.cs

```diff
@@ -0,0 +1,144 @@
+	/// <param name="window">An instance of the <see cref="IWindow"/>.</param>
+	/// <param name="mauiContext">An instance of the <see cref="IMauiContext"/>.</param>
+	/// <remarks>
+	/// Only dragging the top of the titlebar will move the window.
+	/// The top of the TitleBar will also drag the window inside of elements like buttons.
+	/// Gestures such as swiping and controls like swipeview will not work inside the TitleBar.
+	/// </remarks>
+	public WindowViewController(UIViewController windowContentViewController, IWindow window, IMauiContext mauiContext)
+	{
+		_iTitleBarRef = new WeakReference<IView?>(null);
+
+		// Note: Maintain the order for adding a new ViewController to a Container ViewController
+		// 1. Add the Subview
+		// 2. Arrange the Subview's frame
+		// 3. AddChildViewController
+		// 4. Call DidMoveToParentViewController
+
+		if (View is not null && windowContentViewController.View is not null)
+		{
+			_contentWrapperView = new();
+			View.AddSubview(_contentWrapperView);
+			_contentWrapperView.AddSubview(windowContentViewController.View);
+		}
+
+		SetUpTitleBar(window, mauiContext, true);
+		AddChildViewController(windowContentViewController);
+		windowContentViewController.DidMoveToParentViewController(this);
+	}
+
+	public override void ViewWillLayoutSubviews()
+	{
+		base.ViewWillLayoutSubviews();
+		LayoutTitleBar();
+	}
+
+	/// <summary>
+	/// Sets up the TitleBar in the ViewController.
+	/// </summary>
+	/// <param name="window">An instance of the <see cref="IWindow"/>.</param>
+	/// <param name="mauiContext">An instance of the <see cref="IMauiContext"/>.</param>
+	/// <param name="isInitalizing"></param>
+	public void SetUpTitleBar(IWindow window, IMauiContext mauiContext, bool isInitalizing)
+	{
+		var platformWindow = window.Handler?.PlatformView as UIWindow;
+
+		if (platformWindow is null || View is null)
+		{
+			return;
+		}
+
+		var newTitleBar = window.TitleBar?.ToPlatform(mauiContext);
+
+		IView? iTitleBar = null;
+		_iTitleBarRef?.TryGetTarget(out iTitleBar);
+
+		if (newTitleBar != iTitleBar)
+		{
+			_titleBar?.RemoveFromSuperview();
+			iTitleBar = null;
+
+			if (newTitleBar is not null)
+			{
+				iTitleBar = window.TitleBar;
+				View.AddSubview(newTitleBar);
+			}
+
+			_titleBar = newTitleBar;
+			_iTitleBarRef = new WeakReference<IView?>(iTitleBar);
+		}
+
+		var platformTitleBar = platformWindow.WindowScene?.Titlebar;
+
+		if (newTitleBar is not null && platformTitleBar is not null)
+		{
+			platformTitleBar.Toolbar = null;
+			platformTitleBar.TitleVisibility = UITitlebarTitleVisibility.Hidden;
+		}
+
+		LayoutTitleBar();
+
+		// When we are initializing, calling LayoutIfNeeded will cause the layout events to not fire properly.
+		// However we need this when the titlebar is added or removed or the titlebar may not be fully laid out.
+		if (!isInitalizing)
+		{
+			View.LayoutIfNeeded();
```

Comment:
SetNeedsLayout doesn't work?

Reply:
No, it seems that it misses the layout update when using SetNeedsLayout and when a new TitleBar is created after one already existed, the layout is not updated in time until there is another layout pass.

**Using SetNeedsLayout:**

https://github.com/user-attachments/assets/aa2564af-0b98-431f-9f73-9fdc75bad946

**Using LayoutIfNeeded:**

https://github.com/user-attachments/assets/1edc6687-925e-4b14-8dcd-47f5b060925f